### ISSUE TITLE
Complete sockets of an EV charging station

### ISSUE THREAD

**Issue by sideeffect42** [streetcomplete/StreetComplete#5164]
### General
Affected tag(s) to be modified/added: [socket](https://wiki.openstreetmap.org/wiki/Key:socket)
Question asked: **Which sockets does this charging station have (all charging points combined)?**

### Checklist
Checklist for quest suggestions (see [guidelines](https://github.com/streetcomplete/StreetComplete/blob/master/QUEST_GUIDELINES.md)):
- [x] 🚧 To be added tag is established and has a useful purpose
      as an EV driver the locations of charging stations are not worth much if the charging station does not have the appropriate socket for my vehicle, e.g. some electric motocycles or small electric cars can only charge using household sockets while larger vehicles use "special" EV charging connectors.
      In general a charging station without the matching socket is as useful to an EV driver as a filling station only serving petrol is to a driver of a diesel car :-).
- [x] 🤔 Any answer the user can give must have an equivalent tagging (Quest should not reappear to other users when solved by one)
- [x] 🐿️ Easily answerable by any pedestrian from the outside but a survey is necessary
- [x] 💤 Not an overwhelming percentage of quests have the same answer (No spam)
- [x] 🕓 Applies to a reasonable number of map data (Worth the effort)

### Ideas for implementation

**Element selection:** 
Any [amenity:charging_station](https://wiki.openstreetmap.org/wiki/Tag:amenity%3Dcharging_station) which has no `socket`s or where the number of `socket`s combined is less than the station's capacity.

While a charging station can have more sockets available than parking spaces, the other way around does not make much sense :-)

**Proposed UI:**

A list of connectors (with count and maximum charging power). The connector types could be visualised using pictures as it done e.g. for street surfaces.

The maximum charging power is usually printed on the charge point and in most cases voltage/current can be deduced from it, e.g.:
- Type 2 socket with max. 22 kW: 400 V, up to 32 A;
- Type 2 socket with max. 11 kW: 400 V, up to 16 A;
- Type 2 socket with max. 43 A: 400 V, up to 63 A;
- Type 2 socket with max. 7.2 kW: 230 V, up to 32 A;
- Type 2 socket with max. 3.6 kW: 230 V, up to 16 A;
- CCS socket with max. 50 kW: 400 V, up to 125 A;
- CCS socket with max. 150 kW: 400 V, up to 350 A;
- CCS socket with max. 350 kW: 1000 V, up to 350 A.

(However, I don't know how much the voltage/current matters for DC charging (i.e. CCS), because AFAIK when DC charging the EV's battery is connected directly to the charger which will then use the appropriate voltage and current required to charge the battery.)

**Comment by FloEdelmann** [comment]
See https://github.com/streetcomplete/StreetComplete/issues/903#issuecomment-1281157353

**Comment by sideeffect42** [comment]
> See [#903 (comment)](https://github.com/streetcomplete/StreetComplete/issues/903#issuecomment-1281157353)

So what we need is a YAML file with the list of sockets used in which countries.

The types of sockets which can/should be mapped in OSM are [described in the wiki](https://wiki.openstreetmap.org/wiki/Key:socket#List_of_socket_types) and contain all common socket types so that 99% of EV charging stations should be "completable".

By EV charging stations I'm talking about cars, trucks, and larger motocycles; trying to map e-bike recharging stations would be a whole other story since there is no standardisation there at all.

There are some socket types in the OSM wiki which are not used for EV charging (like USB :smile:), so I would limit the list to the socket types which should suffice to map 99% of all charging stations and would suggest to add an option ("has another special socket" or "is not for recharging EVs" which creates an OSM note for somebody with specialised knowledge to map correctly).

Should I try to compile a YAML with countries and sockets used? Would this help in getting this quest to get further?

**Comment by westnordost** [comment]
> Should I try to compile a YAML with countries and sockets used? Would this help in getting this quest to get further?

Yes, absolutely. Also, looking at the socket wiki page, it is really confusing because as you say, sockets for EV, e-bike and "other" (USB-C, really?) charging are all listed in the same table. It would make sense to bring some order in there, too. (But of course, not required for a quest in this app to be realized).

**Comment by georg-d** [comment]
Since most comments of #903, a ground breaking change happened: I learned that since 3/2021, charging points + cars + cables in European Union and European Economic Area carry a symbol "letter on hexagon" to visuablize compatibility. These symbols are well-defined, have standardized appearance, are issued by a relevant authority (CEN), have growing relevance (AFAIK all new products in a whole continent _must_ carry it), have a very simple shape (so are easy to visualize in SC), it's a well-defined list, the list length of 9 allows to re-use existing SC's standard answers GUI look & feel, the symbols are a very easy _and_ reliable way to tell apart different socket types – especially for mappers like me, so not being familiar with electric vehicles and not having practical experience with the plugs – even if they look extremely similar like Type 2 (GB/T 20234.2) and Type 2 (DE, UK). One German source of information on these symbols is https://www.inside-digital.de/ratgeber/buchstaben-an-ladesaeulen-die-raetselhaften-sechseck-symbole 

As a first idea draft: A quest, asking something like "Which symbol(s) appear(s) on the charging point/cable? <reference id>" (the ref is to make it clear which point we're talking of,  similar to quest "levels of a house" showing house number), for answer a multi-selection GUI (like what material recycle containers accept) showing all 9 defined icons and additionaly "Other icon" (creating a note?) + "No icon" (similar to quest asking for bridge weight sign). Of course, we may extend this any time, e.g. by swiping, users may toggle between the symbols and the plug schemes which we talked of so far.

What do you think?

**Comment by matkoniecz** [comment]
> AFAIK all new products in a whole continent must carry it

1) Note "new". Are old ones obligated to get such branding?

2) Is it obligatory to be prominently visible on operational chargers?

3) Is it including say Russia and UK and Switzerland?

> symbol(s)

What is their license? Do you need to pay/license/apply for permission to use them?

**Comment by georg-d** [comment]
1 no, old products are not required to be retro-fitted, so how widespread the symbols are is mostly reflecting the market share of older and newer products at a given point in time

2 yes, to my understanding of the descriptions and what I saw in recent days

3 

- all EU member states
- EEA countries (Iceland, Lichtenstein and Norway)
- North Macedonia, Serbia, Switzerland and Turkey
- Russia is not mentioned in what I did read. By "whole continent" I was referring to the _political_ units – I doubt the _geographically_ European part of Russia will have different rules than the Asian part. Maybe someone knowing Russia could provide information?
- About UK, the leaflet tells _"It is up to the UK to decide the way in which EU rules will apply after Brexit, but newly produced vehicles will have the labels fitted on the production line so vehicles entering the UK market will have the labels."_ Maybe someone knowing UK could provide information on recent political decisions in that matter, or practical/real-life?


Maybe the deep link https://fuel-identifiers.eu/wp-content/uploads/2021/03/EV-labelling_Consumers_EN.pdf works for you? Elsewise, search for directive 2014/94/EU

**Comment by mnalis** [comment]
> 1 no, old products are not required to be retro-fitted, so how widespread the symbols are is mostly reflecting the market share of older and newer products at a given point in time

That one would be quite important for StreetComplete users, I'd think. If it's like 10% chargers with new marks, then it doesn't really help at all (if anything, it would make situation even more complex! e.g. [xkcd 927](https://xkcd.com/927/)). If it is at 90% newly marked, then it would be OK. 

I have not invested a lot of effort into investigation, but in my little part of the EU it would seem it is much more like the former (i.e. I haven't seen a single one marked with those new logos yet. I haven't checked any post-COVID19 charger yet either though; new ones don't exactly pop up every day around here). YMMV.

**Comment by paulklie** [comment]
Btw, I found this collection of icons, that contain the most common domestic sockets: https://www.svgrepo.com/vectors/ev-plug
This project contains more detailed icons for the ev sockets: https://github.com/binary-butterfly/charging-station-connector-icons/tree/bef5b1a22e7b44d338b20983b8f1a56d1e259924
Additionally there are a few with more

**Comment by mcliquid** [comment]
After some research here (mostyl ADAC and EnBW websites) the realistic sockets for car charging stations in the EU:
**AC:**
- type2
- type2_cable
- type1 (very rare, but exists)

**DC:**
- type2_combo (CCS2)
- chademo

**Not to include:**
- tesla_*
- gb_*
- nacs
- type3 (practically obsolete)
- type6/7 (India)
- mcs/j3105 (still extremely rare)

## YAML
Am I correct in assuming that a YAML-file should initially look like this?
```
DE:
  - type2
  - type2_cable
  - type2_combo
  - chademo

AT:
  - type2
  - type2_cable
  - type2_combo
  - chademo

CY:
  - type2
  - type2_cable
  - type2_combo
```

As a first step, I would include EEA + CH + UK at most. Because EEA + UK + CH do not differ significantly from the EU in terms of sockets. Only car-relevant plug types are also easier to cover.

### Basic Quest Idea:

- Multi-select quest (like recycling)
- Only display if:
 - amenity=charging_station
 - no socket:*tags present
- Option “Other type / Unknown” → Note
- Only activate quest if land is included in YAML
- Goal: Maximum of 6–8 options per country

Anything else?

**Comment by matkoniecz** [comment]
You need some clear depictions of different sockets. Maybe photos would work? Or icons?

(hopefully different sockets are distinguishable)

Are all tags to be used are documented and without controversies?

**Comment by mcliquid** [comment]
> You need some clear depictions of different sockets. Maybe photos would work? Or icons?

Both is possible and available.

> Are all tags to be used are documented and without controversies?

Yes

**Comment by matkoniecz** [comment]
I lack expert knowledge here, but it sounds like this is ready to go then!

Hopefully "only part of world is supported" would not be a blocker.

Are you interested in implementing it?

**Comment by mcliquid** [comment]
> I lack expert knowledge here, but it sounds like this is ready to go then!

Yay!

> Hopefully "only part of world is supported" would not be a blocker.

It's definitely much more complex worldwide, but in my opinion it's not impossible. But for the start: the EU is very homogeneous in this respect; we have 3-4 main types that can cover an estimated 80-90% of the infrastructure.
To my knowledge, North America currently has 5 common plugs, with NACS likely to become the new standard.
Asia could be difficult, depending on the country. China, Japan, and India are already very standardized, but I would rule out other countries (also due to a lack of local knowledge). The same goes for South America and Africa.
So, I would stick with the EU, US, Canada, and possibly China and Japan in the long term. The rest would require active community support.

> Are you interested in implementing it?

Yes, I would give it a try at least. I have never worked with YAML files in StreetComplete before, but I enjoy learning new things. To avoid having too many open PRs lying around and making things confusing for everyone, I would wait a few days if necessary.

**Comment by riQQ** [comment]
Sources for icons:
- https://github.com/binary-butterfly/charging-station-connector-icons (CC0)
- https://commons.wikimedia.org/wiki/EV_Charger_Gallery (CC BY-SA 4.0)
- [https://wiki.openstreetmap.org/wiki/Key:socket:*](https://wiki.openstreetmap.org/wiki/Key:socket:*)

But there doesn't seem to be a good visualization to differentiate between Type 2 with and without cable.

**Comment by mnalis** [comment]
>But there doesn't seem to be a good visualization to differentiate between Type 2 with and without cable.

For EEA at least, I'd definitely also include those now-required letter-identifiers https://github.com/streetcomplete/StreetComplete/issues/5164#issuecomment-2045581345 which should make distinction clear

**Comment by riQQ** [comment]
The OSM wiki has the hexagon with letter icons (but only as png) and claims that they are not copyrightable

> These are not eligible for copyright alone because they are not original enough, and thus the image is considered to be in the public domain.


- type2 -> https://wiki.openstreetmap.org/wiki/File:White_C.png
- type2_cable -> https://wiki.openstreetmap.org/wiki/File:Black_C.png
- type2_combo -> https://wiki.openstreetmap.org/wiki/File:Black_K.png, https://wiki.openstreetmap.org/wiki/File:Black_L.png
- chademo -> https://wiki.openstreetmap.org/wiki/File:Black_M.png, https://wiki.openstreetmap.org/wiki/File:Black_N.png

**Comment by westnordost** [comment]
With so many links posted above to SVGs to plug types, without having had a close look at it, I think we can put a checkmark behind finding icons for plug types.

What is missing is the research in which countries which plugs are used.

The quest should likely also include the "household plug" (for e-scooter / e-bike / ... charging) as selection option. It is good that there is a `domestic` value, we can just use that. However, for an adequate picture, I guess we'd still need icons that at least roughly match it. This specifically could be contributed to the countrymetadata repository. A source: https://en.wikipedia.org/wiki/Mains_electricity_by_country#Plugs

Regarding the YAML example @mcliquid posted above: Yes. Although, of course, simplification can take space: `type2` in the metadata could then later expanded into the different types of type2 in case all variations are available (theoretically) in all places where type2 is used.

**Comment by mcliquid** [comment]
> The quest should likely also include the "household plug" (for e-scooter / e-bike / ... charging) as selection option. It is good that there is a `domestic` value, we can just use that.

Due to the complexity and non-standardized abundance of connectors for e-bikes, I would like to completely exclude these connectors from this quest for the time being. There are simply too many different connectors worldwide for this. In my opinion, the `AddChargingStationBicycles` quest should be shown first, followed by the new quest, filtering for `bicycle=no`.

**Comment by mnalis** [comment]
**TL;DR:** I'd go with only **car** chargers matching `amenity=charging_station and bicycle = no and motorcar != no` on the initial version of this Quest, and add (more complete!) bicycle support in subsequent versions.

---

> Due to the complexity and non-standardized abundance of connectors for e-bikes, I would like to completely exclude these connectors from this quest for the time being

I think perhaps you misunderstood? The suggestion was to include **one** generic value only, specifically [socket:domestic=yes](https://wiki.openstreetmap.org/wiki/Key:socket:*). 

But I agree that this suggestion is not ideal. See below on why.

> In my opinion, the `AddChargingStationBicycles` quest should be shown first,  followed by the new quest, filtering for `bicycle=no`.

While I am personally generally more interested in bicycle charging sockets[^1] , this idea does makes most sense to me - make a _"minimal viable product"_ out of the door soon. Trying to get perfect solution in the first go didn't seem to have worked all that well (seeing how long has this and other related issues been around).

IOW I agree with @mcliquid that it is best to **Filter this quest to **only** show few **most popular car chargers** (and only in several easiest countries) matching `amenity=charging_station and bicycle = no and motorcar != no`**

That way, we'll cover many car chargers with **simple** quest initially, and later we can improve the quest:
- react to suggestions from users after using the quest to make it better 
- adding more countries,
- adding more car chargers types,
- but also (more importantly for me) to display **different set of chargers** depending whether the charging station has **`bicycle=yes/no` and/or `motorcar=yes/no`** (because, beside `domestic` charger, there are few more bicycle type chargers like `bosch_3pin`, `bosch_5pin`, `shimano_steps_5pin`, `xlr_3pin_cable`, and [in few countries] `ropd` - which it would be shame not to cover while we're adding bicycle support). 

[^1]:  have to plan in advance for the future when I might not be able to cycle fully manually anymore 🥲

**Comment by mnalis** [comment]
I'll ping @HugoTHOLLON and @NicodeH regarding linked https://github.com/HugoTHOLLON/StreetCompleteSAE_S5/issues/44 -- it seems they were considering or even doing some work on that quest already? Or has that been abandoned? Has it been intended for inclusion in StreetComplete?

**Comment by mcliquid** [comment]
> it seems they were considering or even doing some work on that quest already?

I don't quite understand that either. The ticket has been closed as completed for quite some time, but I can't find any related files in master. There is also no PR for it. I can only find one branch, but as far as I can see, it doesn't contain a quest for sockets: https://github.com/HugoTHOLLON/StreetCompleteSAE_S5/tree/ev_socket

I probably have to wait until my PR has been merged into the countrymetadata before I can even access the YAML content for this quest: https://github.com/streetcomplete/countrymetadata/pull/33

But I'm already working on an icon :)

**Comment by westnordost** [comment]
regarding the work done by HugoTHOLLON's group, I think the project has just run its course. So, I expect no further activity.

> > The quest should likely also include the "household plug" (for e-scooter / e-bike / ... charging) as selection option. It is good that there is a domestic value, we can just use that.

> Due to the complexity and non-standardized abundance of connectors for e-bikes, I would like to completely exclude these connectors from this quest for the time being. There are simply too many different connectors worldwide for this. In my opinion, the AddChargingStationBicycles quest should be shown first, followed by the new quest, filtering for bicycle=no.

`domestic` is for e.g. Schuko-Stecker etc.
I think it should definitely be included. We can still exclude charging stations that are explicitly for bicycles (only) so we don't need to worry about all those special bicycle-charger-sockets, but having domestic sockets is a common feature of normal (i.e. car) charging stations and thus should not be excluded from the quest.

**Comment by mcliquid** [comment]
> `domestic` is for e.g. Schuko-Stecker etc.
> I think it should definitely be included.

I will add them!

**Comment by mcliquid** [comment]
**Quick question regarding socket icons**
According to the EU regulation document, all sockets have a hexagon as a distinguishing feature. I would have placed the plug above it and kept the design very simple. How would you have imagined it?
<img width="335" height="319" alt="Image" src="https://github.com/user-attachments/assets/c65eb385-1480-468e-8eae-0ef2b95d2411" />

**Comment by riQQ** [comment]
Some space between the charger icon and the hexagon would be good to separate them a bit imo.

**Comment by westnordost** [comment]
Don't put the hexagon together with the socket icon into one graphic! 

Note that you can in `override fun ItemContent(item: Socket)` have **any** composable. I.e. you can have a composable like this (simplified; add alignment, padding etc., see e.g. `ImageWithLabel`)

```kt
override fun ItemContent(item: Socket) {
    Column {
      Icon(painterResource(item.icon))
      Icon(painterResource(item.euLabel))
      Text(stringResource(item.name))
    }
}
```

(Using `Icon` over `Image` will also automatically color the icon in "text color", i.e. white in dark mode)

**Comment by HugoTHOLLON** [comment]
> I'll ping [@HugoTHOLLON](https://github.com/HugoTHOLLON) and [@NicodeH](https://github.com/NicodeH) regarding linked [HugoTHOLLON#44](https://github.com/HugoTHOLLON/StreetCompleteSAE_S5/issues/44) -- it seems they were considering or even doing some work on that quest already? Or has that been abandoned? Has it been intended for inclusion in StreetComplete?

Unfortunately our teammate who was working on it gave up I'm sorry. He told me it was too touchy.
